File: leetcode3/DivideIntervalsIntoMinimumNumberofGroups/a.cpp
/* Copyright (C) 2023 by iamslash */

#include <cstdio>
#include <vector>
#include <cmath>

using namespace std;

// intervals: 5,10 6,8 1,5 2,3 1,10
//                                                     i
//      evts: 1,1 1,1 2,1 4,- 6,- 5,1, 6,1 9,- 11,- 11,-
//       cur    0
//       ans:   3

// 609ms 10.46% 146.1MB 19.77%
// sort
// O(NlgN) O(N)
class Solution {
public:
  int minGroups(vector<vector<int>>& intervals) {
    vector<vector<int>> evts;
    for (auto& interval : intervals) {
      evts.push_back({interval[0], 1});
      evts.push_back({interval[1] + 1, -1});
    }
    int ans = 0, cur = 0;
    sort(evts.begin(), evts.end());
    for (auto& evt : evts) {
      ans = max(ans, cur += evt[1]);
    }
    return ans;
  }
};

File: leetcode3/DivideIntervalsIntoMinimumNumberofGroups/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2023 by iamslash

from typing import List

# 1475ms 19.78% 59.2MB 22.53%
# sort
# O(NlgN) O(N)
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        evts = []
        for s, e in intervals:
            evts.append([s, 1])
            evts.append([e + 1, -1])
        ans, cur = 0, 0
        for ts, diff in sorted(evts):
            cur += diff
            ans = max(ans, cur)
        return ans
